# Домашнее задание №3

## Глобальные параметры работы блокнота

In [1]:
# Зафиксируем seed для воспроизводимости
seed = 1

# Выбранная контрольная точка модели
model_checkpoint = "google-t5/t5-small"

# Очистка папки с моделью
! rm -rf ./Model/

## Настройка среды

Импорт необходимых библиотек:

In [2]:
import torch
import numpy as np

from transformers import AutoModel, AutoConfig, AutoTokenizer, AutoModelForSeq2SeqLM, T5ForConditionalGeneration, Seq2SeqTrainingArguments, DataCollatorForSeq2Seq, Seq2SeqTrainer
from datasets import load_dataset, Features, Value, concatenate_datasets, Dataset, DatasetDict
from huggingface_hub import notebook_login
from evaluate import load


Аутентификация в HuggingFace:

In [3]:
notebook_login() # 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

In [4]:
device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
print('{} device is available'.format(device))

if torch.cuda.is_available() and device.type == 'cuda':
    np.random.seed(seed)
     # Если используете CUDA, устанавливаем seed для GPU                
    torch.cuda.manual_seed_all(seed)
    # Для полной воспроизводимости также можно зафиксировать алгоритмы cuDNN
    torch.backends.cudnn.deterministic = True   
    torch.backends.cudnn.benchmark = False
else:
    # Устанавливаем seed для CPU генератора случайных чисел PyTorch
    np.random.seed(seed)
    torch.manual_seed(seed)


# Несколько оптимизаций которые могут ускорить процесс обучения
if torch.backends.cudnn.is_available():
    print(f'Using cudnn version: {torch.backends.cudnn.version()}')
    if device.type == 'cuda':
        torch.backends.cudnn.enabled = True
        torch.backends.cudnn.benchmark = True

cuda device is available
Using cudnn version: 90100


## Задание

[Ссылка](https://tn-gvu.mckx.ru/c/c4YcAACAurcBAEAA/sE43BQ/ZYcLPOYT_IPW6LK3/?u=https%3A%2F%2Fcontest.yandex.ru%2Fcontest%2F67637%2Fenter%2F%3Futm_source%3Dmindbox%26utm_medium%3Demail%26utm_campaign%3Dtraining6%26utm_content%3Ddigest%26utm_term%3D06.11.2024) на контест с заданием.

Необходимо обучить модель перевода с языка зетан на английский.

***Данные:*** https://disk.yandex.ru/d/u8mmcUBn64p_Nw

***Формат решения:*** json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

***Метрика качества:*** BLEU между переводами и английскими референсами на закрытом тесте

## Загрузка данных

In [ ]:
# ! wget https://disk.yandex.ru/d/u8mmcUBn64p_Nw

In [5]:
data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)
dataset

DatasetDict({
    train: Dataset({
        features: ['src', 'dst'],
        num_rows: 300000
    })
    validation: Dataset({
        features: ['src', 'dst'],
        num_rows: 500
    })
    test: Dataset({
        features: ['src', 'dst'],
        num_rows: 1000
    })
})

Посмотрим на один пример данных из каждой части датасета:

In [ ]:
dataset['train'][4]

In [ ]:
dataset['validation'][4]


Немного узнать об этом чудном языке не помешает. Посмотрим сколько уникальных символов в этом языке:

In [6]:
# Функция для подсчета уникальных символов
def count_unique_characters(dataset, feature_name):
    unique_characters = set()
    for split in dataset.keys():
        for example in dataset[split][feature_name]:
            if example is not None:
                unique_characters.update(example)
    return len(unique_characters), unique_characters

# Подсчет уникальных символов во всех частях датасета
num_unique_chars, unique_chars = count_unique_characters(dataset, "src")

print(f"Количество уникальных символов: {num_unique_chars}")
print(f"Уникальные символы: {unique_chars}")

Количество уникальных символов: 182
Уникальные символы: {'ţ', '▸', '¶', 'đ', 'â', '–', '½', '9', '◩', '◂', '’', '™', 'ß', '”', '○', '◡', '◊', '◌', '▢', '◮', '◛', '▱', 'ï', '◈', '♪', '▽', '\u202d', '{', '▾', 'Î', '?', '▼', '5', '◑', '+', ']', '◗', '§', '@', '◇', '▨', '◟', '◖', '△', 'á', '▲', '◓', '◤', 'Ä', '□', 'ú', '◙', '◔', '~', '}', '*', '_', 'Þ', 'ι', '◜', '▬', '°', '\\', '▮', ';', 'û', '◆', '•', '&', '◅', '^', '◞', '◧', 'ë', '\u200b', '▥', ')', 'É', '´', '◭', 'Ý', '▭', 'ο', '2', '\x9d', 'í', '▯', '`', 'ô', '—', 'ν', '◱', '◕', '£', '\x99', '▩', 'ª', ' ', '▹', '▫', '$', '“', '▴', 'é', '=', '¿', '\x9e', ':', '▦', '◠', '[', '8', '◄', 'Â', 'î', '◀', 'å', '◯', '◰', '‚', 'ñ', '◍', '▻', '►', '▵', '◃', '/', '1', 'ó', '◁', '◒', '7', '¤', 'º', '\xa0', '◢', 'ý', '◘', '▿', '◲', '#', '▶', '◚', '▣', 'ä', '%', '●', '¡', '▪', '▷', '◥', '─', '◬', '♫', 'è', '■', '◨', '▤', '6', '"', 'ă', '◪', '◎', '◐', '▧', "'", 'þ', '‘', '3', '◳', '!', '◝', '4', '◦', '0', '◉', '▰', 'ø', '◣', 'ð', '◫', '('}


Посмотрим аналогичную статистику для целевых последовательностей:

In [7]:
# Подсчет уникальных символов во всех частях датасета для целевых последовательностей
num_unique_chars_dst, unique_chars_dst = count_unique_characters(dataset, "dst")

print(f"Количество уникальных символов в целевых последовательностях: {num_unique_chars_dst}")
print(f"Уникальные символы в целевых последовательностях: {unique_chars_dst}")

Количество уникальных символов в целевых последовательностях: 201
Уникальные символы в целевых последовательностях: {'\xad', '¶', 'u', 'U', 'D', 'â', '–', '½', '9', 'Η', 'Y', 'Ã', 'ί', '™', 'x', 'Á', '”', 'g', 'E', 'C', 'p', 'M', 'fl', 'ı', 'Ż', '¬', 'ï', 'H', 'à', 't', '♪', 'Ν', 'v', '\u202d', 'c', 'Z', '{', '˜', 'b', 'æ', '?', 'A', '5', 'R', '+', ']', 'Ë', '§', 'ρ', 'L', '@', 'f', 'á', '│', 'm', 'Ä', 'ú', '~', '}', '*', '_', 'i', 'F', 'o', '\\', '°', '-', 'e', '\x90', 'ì', ';', '&', 'À', 'ç', 'K', 'υ', '^', 'ë', 'ş', ')', 'X', 'É', '€', '´', 'T', 'Ý', '³', '±', 'ο', 'ã', '2', 's', 'í', '\x9d', 'q', '`', 'ô', 'N', '—', 'ν', '£', 'Ö', 'W', '\x99', 'G', 'B', 'ª', '檛', 'Q', ' ', 'у', '鈥', '\x92', '$', '“', 'h', 'é', '=', '¿', '¢', ':', 'š', '[', '8', 'ư', '″', 'Â', 'ü', '.', 'ē', 'Τ', 'å', 'Β', 'l', 'œ', '©', 'Æ', '\x97', 'j', 'k', 'ñ', '‚', 'y', '/', '1', 'V', 'ó', 'Ι', 'Ð', '7', 'r', 'O', '¤', 'fi', 'İ', '¯', 'P', '\xa0', 'ý', 'w', 'õ', '#', 'S', '%', 'ä', 'с', '●', '¡', 'n', '─', '′', '

Посмотрим на максимальную длинну строк для каждой части датасета:

In [8]:
max_lengths = {}

for split in dataset:
    src_lengths = [len(item["src"]) for item in dataset[split] if item["src"] is not None]
    dst_lengths = [len(item["dst"]) for item in dataset[split] if item["dst"] is not None]
    
    max_src_length = max(src_lengths) if src_lengths else 0
    max_dst_length = max(dst_lengths) if dst_lengths else 0
    
    max_lengths[split] = {"src": max_src_length, "dst": max_dst_length}

print(max_lengths)

{'train': {'src': 321, 'dst': 395}, 'validation': {'src': 504, 'dst': 352}, 'test': {'src': 458, 'dst': 0}}


Самая длинная строка присутствующая в датасете имеет длинну 504 символа.

## Подготовка данных

### Токенизатор

#### Нормализация, размер словаря

Нормализация включает в себя некоторую общую очистку, например, удаление ненужных пробельных символов, понижение регистра и/или удаление ударений. Посмотрим как выполняется нормализация токенизатором выбранной нами модели:

In [9]:
example_num = 13

tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

# Проверка нормализации текста
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

print("Модель: ", model_checkpoint)
print("Оригинальный текст:", text)
print("Нормализованный текст:", tokenizer.backend_tokenizer.normalizer.normalize_str(text))
print("Перевод:", translate)

# Проверка, является ли токенизатор быстрым
print("Токенизатор быстрый:", tokenizer.is_fast)

# Узнаем текущий размер словаря токенизатора
current_vocab_size = tokenizer.vocab_size
print(f"Текущий размер словаря токенизатора: {current_vocab_size}")


Модель:  google-t5/t5-small
Оригинальный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Нормализованный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Перевод: Davis, 45, who lives in Leadney, said her son was a great cook and had a charming smile.
Токенизатор быстрый: True
Текущий размер словаря токенизатора: 32100


Судя по выводу текст подвергается минимальной нормализации, значит проблем с потерей символов быть не должно. Размер предобученного словаря токенизатора для выбранной модели равен 32100 токен. 

#### Размер контекстного окна

In [10]:
# Получим максимальную длину последовательности (размер контекстного окна) модели 
max_length = tokenizer.model_max_length

print(f"Максимальная длина последовательности модели: {max_length}")

Максимальная длина последовательности модели: 512


Максимальная длинна контекста для выбранной модели равна 512 токенов.

> ***Длина контекста (или контекстное окно)*** это максимальное количеству токенов, которые модель может обрабатывать одновременно.

#### Обучение нового токенизатора

Так как язык зетан является вымешленным (новым), необходимо обучить новый токенизатор:

In [11]:
def batch_iterator(batch_size=10):
    for split in dataset:
        batch_size = batch_size // 2
        for i in range(0, len(dataset[split]), batch_size):
            src_list = dataset[split][i : i + batch_size]["src"]
            dst_list = dataset[split][i : i + batch_size]["dst"]
            yield [str(src) for src in src_list] + [str(dst) for dst in dst_list]


vocab_size = current_vocab_size  # Можно начать с 32000 и экспериментировать

# Обучение токенизатора
new_tokenizer = tokenizer.train_new_from_iterator(batch_iterator(batch_size=1000), vocab_size=vocab_size, )
print("Токенизатор обучен")




Токенизатор обучен


Проверим работу нового токенизатора на нескольких примерах исходного и целевого текста:

In [12]:
example_num = 1

# Получаем строки из датасета
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

# Токенизируем текст и перевод, выводим результат
print(text)
print(new_tokenizer(text), end="\n\n")

print(translate)
print(new_tokenizer(translate), end="\n\n")

print(new_tokenizer.vocab)

◤◪▦◫ ▨◠▦◞▴◓ ◠◒▪◞▪■ ◀◠◐▪◒◬▨▱▪▨ ◞◫◞▫◪◎◫▦▴ ▨◣▫◭ ▦◫◳◪▫▱◗ ▷▩▼◓◪▱▴◓◫ "◕◣◓◎▴▽◫" ◇◐◓◪▫▴◀◫▱◗◓
{'input_ids': [10746, 18257, 16671, 11032, 14802, 1289, 585, 11063, 1371, 1907, 13501, 11472, 16419, 517, 267, 460, 2572, 7107, 179, 10518, 3414, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

A new cancer vaccine may teach the immune system to "see" abnormal cells
{'input_ids': [251, 543, 17697, 24833, 742, 4487, 107, 22205, 3659, 109, 267, 123, 3219, 179, 111, 565, 472, 10125, 540, 17617, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

{'◳◧◓◞▾▦?': 11320, '◈◈▽': 19746, '▁◫▦◠▦▪◳◂◓': 26318, '▁◀◠▦▨◠': 6364, '▁hoped': 19531, '▁◢◗▱◫◞▴': 30005, '◪▱◫◎': 7319, '▁▭◠': 2556, '▁alien': 7693, '▁▶◠◓◠': 18516, '▁nest': 13241, '▁Massachusetts': 29632, '▁frost': 23912, '▁◞◪◞◞◗▢': 15897, '▁◪◓▨▴◐': 21271, '▁▽◧▱▼◨▱◨▨': 29236, '▁▨â◐◬◈': 31731, '▁cage': 15352, '▁◞◗▢◫▦': 3730, '▁▼◠▦▱◠▦◈▪◓': 19497, "▁else's": 16975, '▁cared': 16597, '▁▶◪◎◗▢'

Сохраним полученный новый токенизатор локально: 

In [13]:
new_tokenizer.save_pretrained("./Model/new_tokenizer")

('./Model/new_tokenizer/tokenizer_config.json',
 './Model/new_tokenizer/special_tokens_map.json',
 './Model/new_tokenizer/tokenizer.json')

### Подготовка данных для обучения модели

Загрузим обученный ранее токенизатор:


In [14]:
# tokenizer = AutoTokenizer.from_pretrained("./Model/new_tokenizer")

tokenizer = AutoTokenizer.from_pretrained("./Model/new_tokenizer")

Проверим токенизатор на паре предложений из тренировочной части датасета:

In [15]:
dataset['train'][example_num]['src']

'▽◪◎◗▦◫▦◫ ▫▴▨◓◠◓ ▴▫◎◪▱◫ ◚▴ ◞◧▦◞▾▢▱◨▨ ◒◠◓◠◀▪▦◈◠▦ ◫◉◎▴▱◫▵'

In [16]:
example_num = 7
test_examples = [dataset['train'][example_num]['src'], dataset['train'][example_num]['dst']]
tokenizer(test_examples)

{'input_ids': [[1057, 474, 3303, 152, 572, 288, 20159, 103, 1], [10068, 3247, 123, 107, 8815, 2009, 147, 1982, 123, 4151, 309, 1]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}

Для разных моделей может понадобиться разный подход в настройке перевода. Например для моделей семества `google-t5` необходим специальный преффикс. Реализуем это:

In [17]:
print(model_checkpoint)

if model_checkpoint in ["google-t5/t5-small", "google-t5/t5-base", "google-t5/t5-larg", "google-t5/t5-3b", "google-t5/t5-11b", "artyomboyko/t5-small-finetuned-zetan-to-en"]:
    print("This is a T5 model")
    prefix = "translate Zetan to English: "
else:
    print("This is not a T5 model")
    prefix = ""

google-t5/t5-small
This is a T5 model


Прежде чем передать в модель тренировочнуй часть датасета необходимо его предварительно обработать. Напишем функцию которая подготовит датасет в которой:

- примеры из датасета будут подаваться на вход `tokenizer` с аргументом `truncation=True`. Это гарантирует, что входные данные, длина которых превышает ту, которую может обработать выбранная модель, будут усечены до максимальной длины, принимаемой моделью.

- Дополнение коротких последовательностей (Padding) будет реализованно позже (в коллаторе данных).

In [18]:
# Установка максимальной длинны входной и выходной последовательности
max_input_length = 512  # max_length
max_target_length = 512 # max_length

source_lang = "zetan"
target_lang = "en"

def preprocess_function(examples):

    inputs = [prefix + ex for ex in examples['src']]
    targets = [ex for ex in examples['dst'] if ex is not None]

    # Токенизируем входные данные и цели
    model_inputs = tokenizer(inputs, max_length=max_input_length, truncation=True)

    with tokenizer.as_target_tokenizer():
        labels = tokenizer(targets, max_length=max_target_length, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

Проверим функцию на нескольких примерах из датасета:

In [19]:
preprocess_function(dataset['train'][:2])

/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:4114: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your labels by using the argument `text_target` of the regular `__call__` method (either in the same call as your input texts if you use the same keyword arguments, or in a separate call.
  warnings.warn(


{'input_ids': [[24970, 9103, 6622, 109, 5326, 355, 10913, 9539, 1284, 3018, 103, 1], [24970, 9103, 6622, 109, 5326, 355, 19465, 1035, 1692, 1514, 261, 163, 1000, 3413, 1817, 24868, 715, 1454, 18114, 1]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], 'labels': [[114, 122, 32007, 104, 1], [213, 243, 240, 109, 7904, 200, 21354, 123, 118, 107, 3804, 115, 107, 1749, 117, 1866, 202, 107, 4572, 115, 18107, 104, 1]]}

Загрузим датасет без тестовой части, так как в ней нет целей. Увеличим валидационную часть датасета, и проведем токенизацию используя метод `map`.

Важно помнить, что увеличение размера валидационной части датасета может оказывать следующее влияние на процесс обучения модели:

***Положительное влияние:***
- Более точная оценка: Большая валидационная выборка обеспечивает более точную оценку производительности модели на невидимых данных. Это помогает лучше понять, насколько хорошо модель обобщает знания и избегает переобучения.
- Лучший выбор гиперпараметров: Больший объем данных для валидации позволяет более надежно сравнивать различные модели и выбирать оптимальные гиперпараметры, что в конечном итоге приводит к лучшей производительности.
- Раннее обнаружение переобучения: С увеличением размера валидационной выборки, переобучение (когда модель хорошо работает на тренировочных данных, но плохо на новых) становится заметнее на ранних стадиях, позволяя быстрее корректировать процесс обучения.

***Отрицательное влияние:***
- Меньше данных для обучения: Увеличение валидационной выборки означает уменьшение объема данных, доступных для обучения. Если общий объем данных ограничен, это может негативно сказаться на производительности модели.
- Увеличение времени обучения: Валидация на большем наборе данных занимает больше времени, что может замедлить процесс обучения.

Нужно найти баланс:
Ключевым моментом является нахождение правильного баланса между размером обучающей и валидационной выборок.
Общие рекомендации: Часто используется соотношение 80/20 или 70/30 для обучающей и валидационной выборок соответственно.



In [20]:

# Стандартное разделение датасета.
data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)
dataset



# # Загрудка датасета с ручным разделением на train и validation
# dataset_seed = 77 #  Уже использовал 42,
# dataset_test_size = 0.1 # Уже использовал 0.01

# features = Features({
#     'src': Value('string'),
#     'dst': Value('string'),
# })

# dataset = load_dataset('json', data_files=['train.jsonl', 'val.jsonl'], features=features)
# dataset = dataset['train'].train_test_split(test_size=0.01, seed=dataset_seed, shuffle=True)
# dataset['validation'] = dataset.pop('test')
# dataset 

DatasetDict({
    train: Dataset({
        features: ['src', 'dst'],
        num_rows: 300000
    })
    validation: Dataset({
        features: ['src', 'dst'],
        num_rows: 500
    })
})

In [21]:
# Применение функции предобработки к датасету
tokenized_datasets = dataset.map(preprocess_function, batched=True, num_proc=20)

tokenized_datasets['train'][0]

Map (num_proc=20):   0%|          | 0/300000 [00:00<?, ? examples/s]

Map (num_proc=20):   0%|          | 0/500 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:4114: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your labels by using the argument `text_target` of the regular `__call__` method (either in the same call as your input texts if you use the same keyword arguments, or in a separate call.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:4114: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your labels by using the argument `text_target` of the regular `__call__` method (either in the same call as your input texts if you use the same keyword arguments, or in a separate call.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:4114: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your 

{'src': '◄▴◓◠▨ ◨▽◠▦◈◬◓▪▼◬▵',
 'dst': '- Intriguing.',
 'input_ids': [24970,
  9103,
  6622,
  109,
  5326,
  355,
  10913,
  9539,
  1284,
  3018,
  103,
  1],
 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
 'labels': [114, 122, 32007, 104, 1]}

# Обучение модели

## Загрузка контрольной точки модели

Загрузим необученную модель (модель со случайно сгенерированными весами):

In [22]:
# Загрузка модели со случайными весами (если хотим обучать с нуля)
# model_checkpoint = "google-t5/t5-small"
# config = AutoConfig.from_pretrained(model_checkpoint)
# model = T5ForConditionalGeneration(config)

# # # Загрузка модели с предобученными весами с Hub для начала обучения
# model_checkpoint = "google-t5/t5-base"
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# # Загрука модели с Hub уже предобученная мной (продолжение обучения)
# model_checkpoint = "artyomboyko/t5-small-finetuned-zetan-to-en"
# model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)
# tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

print("Контрольная точка модели загружена.", model_checkpoint)


Контрольная точка модели загружена. google-t5/t5-small


## Определение метрики

Определим метрику которая будет использоваться для определения качества предсказаний модели в процессе обучения и несколько функций для предобработки результатов:

In [23]:
# Загрузим метрику
metric = load("sacrebleu")

# Функция для постобработки текста
def postprocess_text(preds, labels):
    preds = [pred.strip() for pred in preds]
    labels = [[label.strip()] for label in labels]

    return preds, labels

# Функция для вычисления метрик
def compute_metrics(eval_preds):
    preds, labels = eval_preds
    if isinstance(preds, tuple):
        preds = preds[0]
    decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)

    # Замените -100 в метках, так как мы не можем их декодировать.
    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

    # Немного простой постобработки
    decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)

    result = metric.compute(predictions=decoded_preds, references=decoded_labels)
    result = {"bleu": result["score"]}

    prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in preds]
    result["gen_len"] = np.mean(prediction_lens)
    result = {k: round(v, 4) for k, v in result.items()}
    return result

## Определение аргументов обучения

Определим основные настройки обучения:

При настройках:
batch_size = 112    
lr = 1e-3    
dataset = dataset['train'].train_test_split(test_size=0.01, seed=dataset_seed, shuffle=True)    

Хватит 45 эпох чтобы получить BLEU:    

Лучшая метрика: 18.1667    
Контрольная точка лучшей модели:    

In [ ]:
batch_size = 64 # Оптимальное на текущий момент 64! Изначально было 16, если в VRAM помещается больше, можно увеличить. Это может ускорить обучение и улучшить качество модели.
epochs = 5 # Для подбора lr нужно брать минимум 6 эпох, чтобы стабилизировался лосс и метрика
lr = 2e-3 # Оптимальный на текущий момент от 1e-3 для T5-small (подбором 2e-3). T5-base LR 1e-3 (BLEU 3,00)

model_name = model_checkpoint.split("/")[-1]
args = Seq2SeqTrainingArguments(
    output_dir=f"./Model/{model_name}-finetuned-{source_lang}-to-{target_lang}",
    eval_strategy = "epoch", # Оценка модели в конце каждой эпохи (возможные варианты: "no", "steps", "epoch")
    save_strategy = "epoch", # Сохранение модели в конце каждой эпохи (возможные варианты: "no", "steps", "epoch")
    learning_rate=lr,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    weight_decay=0, # интенсивность L2-регуляризации (изначально было 0.01, при 0.05 - BLEU 3,52) Для T5-base тоже пробовал 0.05 - BLEU .
    save_total_limit=3, # Максимальное количество точек сохранения моделей (одновременно). Чтобы не забивать диск сохранениями на каждой эпохе (или через опеределенное количество шагов)
    num_train_epochs=epochs,
    predict_with_generate=True, # Использование generate для предсказания
    fp16=True,                                                
    push_to_hub=True,   # Публиковать модель на Hugging Face Hub
    hub_private_repo = True, # Использовать для хранения модели приватный репозиторий
    optim = "adamw_torch",  # Оптимизатор
    load_best_model_at_end=True, # Загрузка лучшей модели в конце обучения
    metric_for_best_model="bleu",  # Метрика для выбора лучшей модели
    greater_is_better = True, # Лучшая модель - с наибольшим значением метрики
    report_to=["tensorboard"],  # Отправка метрик в TensorBoard
    warmup_steps=250,   # Количество шагов для warmup
    lr_scheduler_type="linear", # Линейное уменьшение learning rate
    # gradient_accumulation_steps = 2, # Пробуем 2, 
)

## Определение коллатора данных

DataCollator используется для подготовки батчей данных перед подачей их в модель. Основные задачи, которые выполняет DataCollator:
- Создание батчей: Он группирует отдельные примеры данных в батчи, что повышает эффективность обучения и инференса.
- Обработка данных: Может выполнять предварительную обработку данных, например:
- Паддинг: Добавление специальных токенов (padding tokens) к последовательностям разной длины, чтобы сделать их одинаковыми. Это необходимо, так как модели обычно требуют входные данные фиксированной длины.
- Маскирование: Игнорирование определенных токенов во время обучения, например, при реализации masked language modeling.
- Создание дополнительных входных данных: Например, добавление меток классов или идентификаторов.

***Типы DataCollator:***

В Transformers есть несколько предопределенных DataCollator для разных задач:
- `DataCollatorWithPadding`: Динамически добавляет паддинг к последовательностям в батче.
- `DataCollatorForLanguageModeling`: Используется для задач языкового моделирования, применяет маскирование к входным последовательностям.
- `DataCollatorForSeq2Seq`: Для seq2seq задач, таких как машинный перевод.
- `DefaultDataCollator`: Просто объединяет данные в батчи без дополнительной обработки.

Для нашей задачи (перевод) нам нужен специальный коллатор данных (`DataCollatorForSeq2Seq`), который дополнит входные данные и метки до максимальной длины в батче:

In [25]:
data_collator = DataCollatorForSeq2Seq(tokenizer, model=model)

## Обучение модели

Затем нам нужно просто передать все это вместе с нашими датасетами в `Seq2SeqTrainer`:

In [26]:
trainer = Seq2SeqTrainer(
    model,
    args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

/tmp/ipykernel_19467/2282859661.py:1: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Seq2SeqTrainer.__init__`. Use `processing_class` instead.
  trainer = Seq2SeqTrainer(


Теперь мы можем дообучить нашу модель, просто вызвав метод `train`:

In [27]:
trainer.train()


Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.48.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.


Epoch,Training Loss,Validation Loss,Bleu,Gen Len
1,5.283300,6.749310,0.368400,18.626000
2,4.943600,6.466859,0.400000,18.526000
3,4.728300,6.305157,0.582000,18.634000
4,4.595200,6.208886,0.633000,18.580000
5,4.522000,6.171581,0.572600,18.554000


/usr/local/lib/python3.12/dist-packages/transformers/generation/utils.py:1375: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/generation/utils.py:1375: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/generation/utils.py:1375: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/generation/utils.py:1375: UserWarning: Using the model-agnostic default `max_length` (=20) to control

TrainOutput(global_step=23440, training_loss=4.94050396603529, metrics={'train_runtime': 1343.404, 'train_samples_per_second': 1116.567, 'train_steps_per_second': 17.448, 'total_flos': 1.4692206894907392e+16, 'train_loss': 4.94050396603529, 'epoch': 5.0})

Посмотрим на лучшу метрику и лучшую контрольную точку полученные во время обучения:

In [28]:
print(f"Лучшая метрика: {trainer.state.best_metric}")
print(f"Контрольная точка лучшей модели: {trainer.state.best_model_checkpoint}")

Лучшая метрика: 0.633
Контрольная точка лучшей модели: ./Model/t5-small-finetuned-zetan-to-en/checkpoint-18752


После обучения модели будет автоматически загружена лучшая из моделей полученных в процессе обучения, сохраним ее и токенизатор на Hub:

In [ ]:
trainer.push_to_hub()
# model = AutoModel.from_pretrained('./Model/t5-small-finetuned-zetan-to-en/checkpoint-135507')
# model.push_to_hub(repo_id=f"{model_name}-finetuned-{source_lang}-to-{target_lang}")
tokenizer.push_to_hub(repo_id=f"{model_name}-finetuned-{source_lang}-to-{target_lang}")

# Подготовка файла с ответами

Загрузим модель и токенизатор с Hub:

In [ ]:
# translator_checkpoint = "artyomboyko/t5-small-finetuned-zetan-to-en"
# translator_best_checkpoint = "./Model/t5-small-finetuned-zetan-to-en/checkpoint-86728"
model_checkpoint = "artyomboyko/t5-base-finetuned-zetan-to-en"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)
model.to(device)

Формат решения: json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

Загрузим тестовую часть датасета:

In [ ]:
data_files = {
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

test_dataset = load_dataset("json", data_files=data_files, features=features)

print(test_dataset['test'][0])

Для формирования файла ответов необходимо сгенерировать перевод для всех признаков `src` в тестовой части датасета и экспортировать результат в файл json line. Сделаем это следующим образов. Создадим функцию, которая будет принимать на вход пример из датасета и возвращать его перевод. Применим функцию к датасету с помщью функции `map` библиотеки Datasets, после чего сохраним файл:

In [ ]:
def translate_example(example, prefix="translate Zetan to English: "):
    input_text = example["src"]
    input_ids = tokenizer(prefix + input_text, return_tensors="pt").input_ids.to(device)
    outputs = model.generate(input_ids)    
    translated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
    example["dst"] = translated_text
    return example

# Применение функции перевода к тестовому датасету
submission = test_dataset.map(translate_example)

# Сохранение результатов в файл
submission['test'].to_json('submission.jsonl', force_ascii=False)

# Дополнительные материалы

По основной задаче:

- [Training your own tokenizer from scratch](https://github.com/huggingface/notebooks/blob/main/examples/tokenizer_training.ipynb)
- [Translation fine-tuning example](https://github.com/huggingface/notebooks/blob/main/examples/translation.ipynb)

По проблемам с контейнером:

- [Jupyter notebook executions turn grey in Visual studio code](https://stackoverflow.com/questions/73481249/jupyter-notebook-executions-turn-grey-in-visual-studio-code)

# Тестовая часть

In [ ]:
model.push_to_hub(translator_checkpoint)

Используется для отладки кода и проведения экспериментов.